### PR TITLE
Replicated odo table by creating a new devfile properties doc

### PR DESCRIPTION
**Pull request by jc-berger**
Fixes [#312](https://github.com/devfile/api/issues/312).

As a reference, I'm trying to replicate the structure in the [odo devfile properties table](https://odo.dev/file-reference/#devfile-properties).

See the conversation in the merged PR, [#53](https://github.com/devfile/docs/pull/53#pullrequestreview-639288405).

### CODE REVIEW

File: docs/modules/user-guide/partials/proc_devfile-properties.adoc [hunk 1]
```diff
@@ -0,0 +1,258 @@
+[id="proc_devfile-properties_{context}"]
+= Devfile properties
+
+[role="_abstract"]
+This page quickly introduces you to devfile properties. The code snippets help you get started developing your own devfile projects. The tables show you what the different devfile properties require if you use these properties. See the following table of contents to go to the devfile property you want to work with:
```
Comment (jc-berger) [inline docs/modules/user-guide/partials/proc_devfile-properties.adoc:5]: Think I should say "users" here since it's not a procedure, so I can't directly address users? I want to make it user focus, but I don't want to assume the user is doing or thinking anything particular by saying 'you` here.

File: docs/modules/user-guide/nav.adoc [hunk 2]
```diff
@@ -1,4 +1,5 @@
 * xref:index.adoc[]
+* xref:devfile-properties.adoc[]
```
Comment (jc-berger) [inline docs/modules/user-guide/nav.adoc:2]: How does the placement of this new doc look in the TOC?

File: docs/modules/user-guide/partials/proc_devfile-properties.adoc [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+|eventObject
+
+|Required
+|no
+
+|Description
+|events to be executed during a project lifecycle
+|===
+
+This events example uses an Open Liberty container. For more information about the events in devfiles, see: xref:adding-event-bindings.adoc[].
```
Comment (jc-berger) [inline docs/modules/user-guide/partials/proc_devfile-properties.adoc:258]: I'm thinking I'll add an "additional resources" section at the end to direct users to other devfile docs to continue their work on devfiles. Thoughts?

File: docs/modules/user-guide/partials/proc_devfile-properties.adoc (hunk 1 above)
Comment (elsony) [reply docs/modules/user-guide/partials/proc_devfile-properties.adoc:5]: I think using "you" is fine in here

File: docs/modules/user-guide/nav.adoc (hunk 2 above)
Comment (elsony) [reply docs/modules/user-guide/nav.adoc:2]: The placement is ok but I wonder if the title should be something like "Devfile quick reference" (trying to differentiate this between the spec/schema link and the stack author's guide.

File: docs/modules/user-guide/partials/proc_devfile-properties.adoc (hunk 3 above)
Comment (elsony) [reply docs/modules/user-guide/partials/proc_devfile-properties.adoc:258]: Adding an additional resources link is good if we position this one as a quick reference.

File: docs/modules/user-guide/partials/proc_devfile-properties.adoc [hunk 4]
```diff
@@ -0,0 +1,258 @@
+[id="proc_devfile-properties_{context}"]
+= Devfile properties
+
+[role="_abstract"]
+This page quickly introduces you to devfile properties. The code snippets help you get started developing your own devfile projects. The tables show you what the different devfile properties require if you use these properties. See the following table of contents to go to the devfile property you want to work with:
+
```
Comment (elsony) [inline docs/modules/user-guide/partials/proc_devfile-properties.adoc:6]: I wonder if there is a way to make it more concise. Comparing to the original version from the odo doc, that one looks much more like a quick reference guide than this. In order to find the information that I need, I'll need to scroll through the elements here. @kadel what do you think?

**Review by elsony — CHANGES_REQUESTED** [review]

File: docs/modules/user-guide/partials/proc_devfile-properties.adoc (hunk 4 above)
Comment (kadel) [reply docs/modules/user-guide/partials/proc_devfile-properties.adoc:6]: Yeh, in the current format it harder to read :-/ 
The big advantage of how it is displayed in odo.dev is that the example is on the definitions and examples are side-by-side.

Generating html version of this in a similar format like we have in https://odo.dev/file-reference/#devfile-properties would help with readability a lot. 

Just for reference markdown source for that page is at https://github.com/openshift/odo/blob/gh-pages/slate/source/index.html.md

Comment (jc-berger) [reply docs/modules/user-guide/partials/proc_devfile-properties.adoc:6]: Thanks Elson and Tomas. Not sure if you all are able to see the rendered output, so I included some screenshots. Yes, this doc I wrote doesn't have the 3 sections like the odo table...I'm not sure if we're even able to do that in Asciidoc. I've reached out to more writers on my team and will do more research on converting Markdown to Asciidoc. I'll keep everyone posted on how much Markdown capabilities we can replicate and utilize in Asciidoc. 
SS of rendered odo table replica
2nd SS of rendered odo table replica

File: docs/modules/user-guide/partials/proc_devfile-quick-reference.adoc [hunk 5]
```diff
@@ -0,0 +1,263 @@
+[id="proc_devfile-quick-reference_{context}"]
+= Devfile quick reference
+
+[role="_abstract"]
+This page quickly introduces you to devfile properties. The code snippets help you get started developing your own devfile projects. The tables show you what the different devfile properties require if you use these properties. Refer to the table of contents to go to the devfile property you want to work with.
```
Comment (ossontoes) [inline docs/modules/user-guide/partials/proc_devfile-quick-reference.adoc:5]: As before, I'm not sure what's an expected language style in this set of docs, maybe a more descriptive and "user welcoming" one is preferred.
On "my" projects writers would avoid describing what users are expected to know (e.g. that code snippets contain code examples or that users can use a table of contents). Users tend to scroll down to code snippets, and the faster they get info, the more efficient our docs are considered.

In the end, our reviewers would cut this intro paragraph to something like:
This page describes devfile properties and requirements for using them.

File: docs/modules/user-guide/partials/proc_devfile-quick-reference.adoc [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+
+== SchemaVersion
```
Comment (ossontoes) [inline docs/modules/user-guide/partials/proc_devfile-quick-reference.adoc:7]: It looks like schemaVersion is a proper name, perhaps it should begin with a lowercase letter even in a title, WDYT? E.g. eBay wouldn't be spelled EBay in a title. But I haven't managed to find a relevant rule about that in the IBM SG.

File: docs/modules/user-guide/partials/proc_devfile-quick-reference.adoc [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+schemaVersion: 2.0.0
+----
+
+[cols="1,1"]
+|===
+|Key
+|schemaVersion
+
+|Type
+|string
```
Comment (ossontoes) [inline docs/modules/user-guide/partials/proc_devfile-quick-reference.adoc:22]: Type String is capitalized in some languages, e.g. in Java.
Sorry, I lack context, maybe it's not important here.

File: docs/modules/user-guide/partials/proc_devfile-quick-reference.adoc [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+|starterProjectObject
+
+|Required
+|no
+
+|Description
+|list of starter projects to be used
+|===
+
+This starterProject example uses a Quarkus framework. For more information about the starterProjects in devfiles, see: xref:migrating-projects.adoc[].
```
Comment (ossontoes) [inline docs/modules/user-guide/partials/proc_devfile-quick-reference.adoc:92]: the Quarkus framework? Or simply Quarkus maybe?

File: docs/modules/user-guide/partials/proc_devfile-quick-reference.adoc [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+|componentObject
+
+|Required
+|no
+
+|Description
+|list of components to be used
+|===
+
+This component example uses an Open Liberty container, For more information about the components in devfiles, see: xref:migrating-components.adoc[].
```
Comment (ossontoes) [inline docs/modules/user-guide/partials/proc_devfile-quick-reference.adoc:140]: ...container. (period)

File: docs/modules/user-guide/partials/proc_devfile-quick-reference.adoc [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+|Required
+|no
+
+|Description
+|events to be executed during a project lifecycle
+|===
+
+This events example uses an Open Liberty container. For more information about the events in devfiles, see: xref:adding-event-bindings.adoc[].
+
+== Additional Resources
```
Comment (ossontoes) [inline docs/modules/user-guide/partials/proc_devfile-quick-reference.adoc:260]: Red Hat projects switched to sentence case for titles (Additional resources), but maybe it doesn't apply to this project?

File: docs/modules/user-guide/partials/proc_devfile-quick-reference.adoc (hunk 5 above)
Comment (jc-berger) [reply docs/modules/user-guide/partials/proc_devfile-quick-reference.adoc:5]: Good suggestion--nice and concise. I've made the change.

File: docs/modules/user-guide/partials/proc_devfile-quick-reference.adoc (hunk 6 above)
Comment (jc-berger) [reply docs/modules/user-guide/partials/proc_devfile-quick-reference.adoc:7]: You're right. Proper names and nouns are always spelt the same way no matter where they are in headings or sentences.

File: docs/modules/user-guide/partials/proc_devfile-quick-reference.adoc (hunk 7 above)
Comment (jc-berger) [reply docs/modules/user-guide/partials/proc_devfile-quick-reference.adoc:22]: Good point. I'll leave as is (un-capitalized) because it's not capitalized on the odo page where we are taking this content.

File: docs/modules/user-guide/partials/proc_devfile-quick-reference.adoc (hunk 10 above)
Comment (jc-berger) [reply docs/modules/user-guide/partials/proc_devfile-quick-reference.adoc:260]: Yes, once again, good catch, thank you! I've made the change!

File: docs/modules/user-guide/partials/proc_devfile-properties.adoc (hunk 4 above)
Comment (jc-berger) [reply docs/modules/user-guide/partials/proc_devfile-properties.adoc]: @elsony and @kadel I was talking to the other writers about Asciidoc and Antora capabilities. Using Antora to build our docs, we don't have the ability to replicate the three pane view that's used in the odo page. I was also recommended to use lists instead of tables....tables are harder to edit, build, and use for accessibility purposes. I've included some screenshots of the rendered version. As you can see, the horizontal lists mirrors the table pretty well. 

Let me know what you guys think, thanks! 
Screen Shot 2021-06-07 at 3 22 17 PM
Screen Shot 2021-06-07 at 3 22 27 PM
Screen Shot 2021-06-07 at 3 22 34 PM

**Comment by missmesss** [comment]
Sorry for the ghost account mess...

The changes LGTM.

The only last thing: following the schemaVersion / SchemaVersion situation, I can't stop looking at the metadata and starterProjects cases. Will the same apply there?

And if starterProjects is a set property name, should is keep its plural form in this sentence:
This starterProject example uses Quarkus?

**Comment by jc-berger** [comment]
@missmesss sorry for the late reply. Just seeing this :) 

So, as far as capitalization, I'll keep it, `starterProjects, metedata, metadaObject, starterProjectObject, and commandObject` because that's how the proper names are spelt from the odo page (where I'm taking the bulk of this content). However, you're right on the money with the case, starterProject**s**. I'll make sure it keeps its plural form throughout the doc. Making a new commit now. Thanks!

**Comment by missmesss** [comment]
Thank you for clarifying this, @jc-berger.
If they're components' proper names that always start with lowercase letters, then shouldn't the titles be that way too (like schemaVersion)?

 == schemaVersion
 == Metadata
 == StarterProjects

**Comment by jc-berger** [comment]
@missmesss , I see. Yes, you're right. Good catch. I'll correct the capitalization. Thank you! :) 

Other than, LGTM, right?

**Comment by missmesss** [comment]
Yes, Jake, docs look good, thanks for all the hard work.

**Review by missmesss — APPROVED** [review]

File: docs/modules/user-guide/partials/proc_devfile-properties.adoc (hunk 4 above)
Comment (missmesss) [reply docs/modules/user-guide/partials/proc_devfile-properties.adoc]: I'm sorry to jump in @elsony and @kadel, writers aren't big web design specialists, ofc, but there're some related topics that we discuss and research, e.g. user experience and accessibility.

It can be just me, but I don't see how this wide and packed 3-pane layout https://odo.dev/file-reference/#devfile-properties is easy to use, sorry.

For me the 2nd and 3rd panes sometimes get an extra scroll bar at the cell bottom, information gets hidden, it means extra clicking and scrolling, more time to get info, not user-friendly at all.
If we zoom out to see all content, it'll be very wide, meaning more eye movement and more time spent on a page again. IIRC, UI/UX designers' work is to provide info faster and make users click less.

Accessibility-wise, how can we be sure that all screen readers correctly connect code examples with devfile properties? Tables are a big accessibility concern on their own, and these 3 panes with tables are potentially very confusing to screen readers.

In short, @jc-berger 's proposed design might make a page longer, but in the end it provides information faster and in a more straightforward and accessible way. Red Hat docs writers are currently trained to choose simpler and more lightweight structures, hence Jake's proposal, I guess.

Comment (kadel) [reply docs/modules/user-guide/partials/proc_devfile-properties.adoc]: > It can be just me, but I don't see how this wide and packed 3-pane layout https://odo.dev/file-reference/#devfile-properties is easy to use, sorry.

The feedback that we got on this layout was only positive.

**Comment by kadel** [comment]
I'm sorry but currently, this looks like an incomplete version of https://docs.devfile.io/devfile/2.1.0/user-guide/api-reference.html

I thought that this is work supposed to enrich the existing [API reference ](https://docs.devfile.io/devfile/2.1.0/user-guide/api-reference.html) to make it easier to consume. 
Maybe I'm missing something but I'm not sure how this document in its current form helps anyone to make sense of the devfile structure :-(

**Comment by jc-berger** [comment]
Thank you both @missmesss and @kadel for your reviews and thoughtful comments :) 

@kadel, yes, I agree, that the doc in this PR doesn't improve the user experience with our docs. I'm thinking our current Authoring Stacks has all the content from the odo table, and in addition, we have user stories, procedures, and abstracts. 

Since we pretty much have the odo table content throughout our authoring stacks, can we focus on improving those authoring stacks docs? For example, in addition to the abstract for each authoring stack doc, we can include a list, like the odo table has, saying the `key`, `type`, `description`. 

I'm thinking that focusing on the authoring stacks is the best way to improve the content and user experience. Thoughts?

**Comment by jc-berger** [comment]
Per our conversations, I've included the odo content throughout the authoring stacks. Since this is a new approach, I've opened a new PR. To review the odo content throughout the authoring stacks, please see PR #78: https://github.com/devfile/docs/pull/78

**Comment by elsony** [comment]
@ dharmit Here is the PR for the odo doc migration that we mentioned on the odo call earlier.

**Comment by kadel** [comment]
> @ dharmit Here is the PR for the odo doc migration that we mentioned on the odo call earlier.

@dharmit :wink:

**Comment by jc-berger** [comment]
For the most recent work on the odo and devfile consolidation, please see PR 78: https://github.com/devfile/docs/pull/78

**Comment by jc-berger** [comment]
Closing this PR as this devfile and odo consolidation work has now been addrssed in the new PR: https://github.com/devfile/docs/pull/78
